ophyd/v2/core.py

```diff
@@ -854,74 +802,68 @@ async def merge_gathered_dicts(
+            Signals to make up `read()`
         conf:
-            Signals to make up `read_configuration()` that can be cached
+            Signals to make up `read_configuration()`
+        read_uncached:
+            Signals to make up `read()` that won't be cached
         """
-        self._read_signals = tuple(read) + tuple(
-            _UncachedSignal(sig) for sig in read_uncached
-        )
-        if primary:
-            self._read_signals = (_ReadableRenamer(primary, self),) + self._read_signals
-        self._conf_signals = tuple(config)
-        self._staged = False
-        # Call this last so child Signals are renamed
-        if name:
-            self.set_name(name)
-
-    def stage(self) -> List[Any]:
-        self._staged = True
-        staged = [self]
-        for sig in self._read_signals + self._conf_signals:
-            if isinstance(sig, Stageable):
-                staged += sig.stage()
-        return staged
-
-    def unstage(self) -> List[Any]:
-        self._staged = False
-        unstaged = [self]
-        for sig in self._read_signals + self._conf_signals:
-            if isinstance(sig, Stageable):
-                unstaged += sig.unstage()
-        return unstaged
+        self._read_signals = tuple(read)
+        self._configuration_signals = tuple(config)
+        self._read_uncached_signals = tuple(read_uncached)
 
-    async def describe(self) -> Dict[str, Descriptor]:
-        return await merge_gathered_dicts(sig.describe() for sig in self._read_signals)
+    @AsyncStatus.wrap
+    async def stage(self) -> None:
+        for sig in self._read_signals + self._configuration_signals:
+            await sig.stage().task
 
-    async def read(self) -> Dict[str, Reading]:
-        return await merge_gathered_dicts(sig.read() for sig in self._read_signals)
+    @AsyncStatus.wrap
+    async def unstage(self) -> None:
+        for sig in self._read_signals + self._configuration_signals:
+            await sig.unstage().task
 
     async def describe_configuration(self) -> Dict[str, Descriptor]:
-        return await merge_gathered_dicts(sig.describe() for sig in self._conf_signals)
+        return await merge_gathered_dicts(
+            [sig.describe() for sig in self._configuration_signals]
+        )
 
     async def read_configuration(self) -> Dict[str, Reading]:
-        return await merge_gathered_dicts(sig.read() for sig in self._conf_signals)
+        return await merge_gathered_dicts(
+            [sig.read() for sig in self._configuration_signals]
+        )
+
+    async def describe(self) -> Dict[str, Descriptor]:
+        return await merge_gathered_dicts(
+            [sig.describe() for sig in self._read_signals + self._read_uncached_signals]
+        )
+
+    async def read(self) -> Dict[str, Reading]:
+        return await merge_gathered_dicts(
+            [sig.read() for sig in self._read_signals]
+            + [sig.read(cached=False) for sig in self._read_uncached_signals]
+        )
```

Comment:
Let's assume you need to set pv1 to a value, and when that completes you can read pv2. If you use cached signals this will be equivalent to:

```python
values = []
camonitor(pv1, values.append)
caput(pv2, wait=True)
return values[-1]
```

However there is a race condition between the caput callback coming in, and the monitor update coming in. To be safe, we should use uncached signals for these, which is less timely but guaranteed to be correct:

```python
caput(pv2, wait=True)
return caget(pv1)
```